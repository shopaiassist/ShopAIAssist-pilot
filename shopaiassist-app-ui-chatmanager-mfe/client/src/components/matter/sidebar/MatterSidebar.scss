@import "../../../utils/styles/imports";

$headerHeight: 3.75rem;

.matter-sidebar-container {
   display: flex;
   width: 21rem;
   z-index: map-get($z-index-map, "default");

   &.compact-mode {
      position:absolute;
      width: 100%;
   }

   &.full-mode {
      position:relative;
      width: 21rem;
      max-width: 100%;
   }

   .matter-sidebar-nav-header {
      display: flex;
      padding: 0.5rem;
      justify-content: space-between;
      align-items: center;
      align-self: stretch;
      min-height: 3.688rem;
      box-sizing: border-box;
      width: 100%;
      border: rem-calc(1);
      border-bottom: $saf-line-width-thin solid $saf-color-line-subtle;
      border-left: $saf-line-width-thin solid $saf-color-line-subtle;

      .matter-sidebar-nav-back {
         display: flex;
         gap: 0.5rem;
      }
   }

   .matter-sidebar {
      display: flex;
      flex-direction: column;
      align-items: center;
      box-shadow: $saf-drop-shadow-level-3;
      height: calc(100vh - $headerHeight);
      width: 21rem;
      background-color: $saf-color-background-subtle;

      &.compact-mode {
         width: 100%;
      }

      &.full-mode {
         width: 21rem;
         max-width: 100%;
      }

      .matter-sidebar-folder-name {
         padding: 0.5rem;
         margin: 0.75rem;
         display: flex;
         align-self: flex-start;
      }

      .matter-sidebar-header {
         display: flex;
         flex-direction: row;
         justify-content: space-between;
         width: 95%;
         padding: 0.5rem;
      }

      .matter-sidebar-footer {
         display: flex;
         width: 100%;
      }
   }
}