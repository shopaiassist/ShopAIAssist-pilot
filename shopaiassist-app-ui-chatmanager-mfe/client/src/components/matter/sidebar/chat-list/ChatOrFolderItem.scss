@import "../../../../utils/styles/imports";

.chat-or-folder-item-container {
   display: flex;
   flex-direction: row;
   align-items: center;
   row-gap: 1.5rem;
   column-gap: 1rem;
   padding: 0.25rem;
   justify-content: center;

   .chat-or-folder-item-menu {
      margin-left: auto;
      margin-right: 1rem;
   }

   .chat-or-folder-menu {
      position: relative;

      .chat-or-folder-saf-menu {
         display: flex;
         flex-direction: column;
         position: fixed;
         width: fit-content;
         z-index: map-get($z-index-map, "default");
      }
   }

   .chat-or-folder-icon::part(fa-icon) {
      padding-top: 0.15rem;
   }
}
