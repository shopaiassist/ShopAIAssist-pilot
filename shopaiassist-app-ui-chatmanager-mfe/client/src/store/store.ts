import { Action, AsyncThunkPayloadCreator, configureStore, ThunkAction } from '@reduxjs/toolkit';

import userReducer from './userSlice';

export const reducer = {
  user: userReducer
};

export const store = configureStore({
  reducer
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;

export interface AppThunkApiConfig {
  dispatch: AppDispatch;
  rejectValue: string;
  state: RootState;
}
export type AppAsyncThunk<P, R> = AsyncThunkPayloadCreator<R, P, AppThunkApiConfig>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
