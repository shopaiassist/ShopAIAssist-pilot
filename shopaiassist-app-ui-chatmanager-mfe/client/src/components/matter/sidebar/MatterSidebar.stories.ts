// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
import MatterSidebar from './MatterSidebar';
import { Meta, StoryObj } from '@storybook/react';

const meta = {
  title: 'Matter/Sidebar',
  component: MatterSidebar,
  parameters: {
    layout: 'padded'
  },
  tags: ['autodocs']
} satisfies Meta<typeof MatterSidebar>;

export default meta;

export const Default: StoryObj = { args: {} };
