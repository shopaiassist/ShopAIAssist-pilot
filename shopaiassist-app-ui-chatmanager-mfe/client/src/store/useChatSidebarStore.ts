import { create } from 'zustand';
import { devtools } from 'zustand/middleware';
import axios from 'axios';
import { debounce } from 'lodash';
import { FolderItem, TreeItem } from '../@types/sidebar';
import { Endpoints } from '../utils/api';
import { UserAuth } from 'react';
import { useMfeContextUtilsStore } from './useMfeContextUtilsStore';
import i18nInstance from '../utils/i18n';
import { useBreadCrumbStore } from './useBreadcrumbStore';
import { GENERATE_CHAT_NAME_WAIT_TIME } from '../utils/constatnts';
import { UserProfile } from 'react/dist/types/auth';

export enum SortOptions {
  BY_DATE = 'by_date',
  BY_NAME = 'by_name',
  BY_TYPE = 'by_type'
}

export const ITEM_TYPE = {
  CHAT: 'chat',
  FOLDER: 'folder'
} as const;

/**
 * Defines the structure and types of the state managed in the ChatSidebarStore.
 * @interface ChatSidebarState
 * @property {UserAuth|undefined} userAuth - The user authentication state.
 * @property {(TreeItem|FolderItem)[]|undefined} chatsAndFolders - The list of chats and folders.
 * @property {string} sortType - The current sorting type for displaying the chats and folders.
 * @property {FolderItem|undefined} activeFolder - The currently active folder.
 * @property {TreeItem|undefined} activeChat - The currently active chat.
 * @property {Function} fetchChatList - Function to fetch the list of chats and folders based on the current sort type.
 * @property {Function} deleteChat - Function to delete a chat by its identifier.
 * @property {Function} createChat - Function to create a new chat.
 * @property {Function} updateChatName - Function to update the name of a chat.
 * @property {Function} setSortType - Function to update the sorting type and fetch the updated list.
 * @property {Function} setActiveFolder - Function to set the active folder and fetch its chats.
 * @property {Function} setActiveChat - Function to set the active chat.
 * @property {Function} setUserAuth - Function to set the user authentication state.
 */
export interface ChatSidebarState {
  userAuth: UserAuth | undefined;
  chatsAndFolders?: (TreeItem | FolderItem)[];
  sortType: string;
  activeFolder?: FolderItem;
  activeChat?: TreeItem;
  archivedMattersView?: boolean;
  isNameUpdate: boolean;
  isMatterSettingsActive?: boolean;
  isMatterFilesActive: boolean;
  isLoading: boolean;
  activeUser: UserProfile;
  fetchChatList: () => void;
  deleteChat: (chatId: string) => void;
  createChat: (message: string, chatFolderIdentifier?: string) => Promise<string>;
  updateChatName: (chatId: string, newChatName: string) => void;
  generateChatName: (chatId: string, chatHistory: { role: string, content: string | undefined }[]) => void;
  setSortType: (newSortType: string) => void;
  setActiveFolder: (newActiveFolder?: FolderItem) => void;
  setActiveChat: (newActiveChat?: TreeItem) => void;
  setUserAuth: (userAuth: UserAuth) => void;
  toggleArchivedMattersView: () => void;
  setIsMattersettingsActive: (isMatterSettingsActive: boolean) => void;
  refreshActiveFolder: () => Promise<void>;
  setIsMatterFilesActive: (isMatterFilesActive: boolean) => void;
  setIsNameUpdate: (isNameUpdateNewValue: boolean) => void;
  setActiveUser: (activeUser: UserProfile) => void;
}

/**
 * The base URL constructor for API endpoints, utilizes the environment variable for the app's domain.
 * @param {string} endpoint - The API endpoint to be appended to the base domain.
 * @returns {string} The full URL for the API call.
 */
const BASE_URL = (endpoint: string) => `${process.env.APP_DOMAIN}/api/${endpoint}`;

// const SAMPLE_CHATS: TreeItem[] = [
//   {
//     treeItemId: 'c91fed15-2e50-45eb-a59a-666666667777',
//     name: '--A - CREATE NEW CHAT--',
//     createdAt: new Date('2024-03-04T18:43:32.194686+00:00'),
//     updatedAt: new Date(),
//     type: 'chat',
//     fileCollectionId: ''
//   },
//   {
//     treeItemId: 'c91fed15-2e50-45eb-a59a-cebe44e2bf7b',
//     name: 'ZZZConversation Naming Assistant Capabilities',
//     createdAt: new Date('2024-03-04T18:43:32.194686+00:00'),
//     updatedAt: new Date('2024-03-04T18:43:32.194686+00:00'),
//     type: 'chat',
//     fileCollectionId: ''
//   },
//   {
//     treeItemId: '5a44626f-b1b8-4943-acc3-62651e2c39a3',
//     name: 'PDF File Upload: 070823vanliere (1)',
//     createdAt: new Date('2024-03-05T16:32:52.361758+00:00'),
//     updatedAt: new Date('2024-03-05T16:32:52.361758+00:00'),
//     type: 'chat',
//     fileCollectionId: ''
//   }
// ];

/**
 * Zustand store to manage the sidebar state, including chats and folders and related actions.
 * Uses axios for HTTP requests and lodash's debounce for limiting the rate of API calls.
 */
export const useChatSidebarStore = create<ChatSidebarState>()(
  devtools(
    (set, get) => ({
      userAuth: undefined,
      chatsAndFolders: [],
      sortType: SortOptions.BY_DATE,
      activeFolder: undefined,
      activeChat: undefined,
      isMatterSettingsActive: undefined,
      isMatterFilesActive: false,
      isNameUpdate: false,
      isLoading: false,
      activeUser: {
        email: '',
        firstName: '',
        lastName: '',
        username: '',
        registrationKey: '',
        region: '',
        id: ''
      },
      /**
       * Fetches the list of chats and folders based on the current sort type.
       * Uses a debounced function to limit the rate of API calls.
       * Sets isLoading to true during the fetch operation.
       */
      fetchChatList: debounce(async () => {
        // Set loading state to true when starting fetch
        set({ isLoading: true });
        try {
          const res = await axios.get(BASE_URL(Endpoints.LIST), {
            params: {
              sortType: get().sortType,
              parentId: get().activeFolder?.treeItemId,
              onlyArchivedMatters: get().archivedMattersView,
              uid: get().activeUser.id
            }
          });
          // Validate API response and update state with new data
          const chatsAndFolders = Array.isArray(res.data) ? res.data : [];
          if (!Array.isArray(res.data)) {
            console.error('API returned unexpected data structure for fetchChatList:', res.data);
          }
          set({ chatsAndFolders, isLoading: false });
        } catch (error) {
          // Make sure to set loading to false even if there's an error
          console.error('Error fetching chat list:', error);
          set({ isLoading: false });
          useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
        }
      }),
      /**
       * Deletes a chat by its identifier and refreshes the chat list.
       * Uses a debounced function to limit the rate of API calls.
       *
       * @param {string} chatId - The identifier of the chat to be deleted.
       */
      deleteChat: debounce(async (chatId: string) => {
        const deleteChatResult = await axios.delete(BASE_URL(Endpoints.CHAT_BY_ID), {
          data: { id: chatId, user: get().activeUser }
        });
        if (deleteChatResult.status === 200) {
          useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('CHAT_DELETE_SUCCESS'), 'success');
        } else {
          useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
        }
        get().setActiveChat();
        get().fetchChatList();
      }),

      /**
       * Creates a new chat and returns the chat ID.
       *
       * @param {string} [chatFolderIdentifier] - The identifier of the chat folder.
       * @returns {Promise<string>} The new chat ID.
       */
      createChat: async (message: string | undefined, chatFolderIdentifier?: string): Promise<string> => {
        const newChatData = await axios.post(
          BASE_URL(Endpoints.LIST),
          {
            parentId: chatFolderIdentifier,
            user: get().activeUser
          },
          { withCredentials: true }
        );

        get().setActiveChat(newChatData.data.newChat as TreeItem);
        console.info(`Created new chat with ID: ${newChatData.data.newChat.treeItemId}`);
        get().generateChatName(newChatData.data.newChat.treeItemId, [{ role: 'user', content: message ?? ''}] );
        return newChatData.data.newChat.treeItemId as string;
      },
      /**
       * Generates a new name for a chat and updates the chat name.
       * Uses a debounced function to limit the rate of API calls.
       *
       * @param {string} chatId - The identifier of the chat to be updated.
       */
      generateChatName: debounce(async (chatId: string, chatHistory: { role: string, content: string | undefined }[]) => {
        setTimeout(async () => {
          try {
            const generatedName = await axios.post(BASE_URL('generate-name'), { chatId, chat_history: chatHistory });

            if (generatedName.data.newName) {
              const activeChat = get().activeChat;

              if (activeChat !== undefined && activeChat.treeItemId === chatId) {
                set({ isNameUpdate: true });
                get().setActiveChat({ ...activeChat, name: generatedName.data.newName });
                
                // wait for a moment then update the chat list.
                setTimeout(() => {
                  get().fetchChatList();
                }, GENERATE_CHAT_NAME_WAIT_TIME);
              }
            }
          } catch (error) {
            console.error('Error generating chat name:', error);
            useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
          }          
        }, GENERATE_CHAT_NAME_WAIT_TIME);
      }),
      /**
       * Updates the name of a chat and refreshes the chat list.
       * Uses a debounced function to limit the rate of API calls.
       *
       * @param {string} chatId - The identifier of the chat to be updated.
       * @param {string} newChatName - The new name for the chat.
       */
      updateChatName: debounce(async (chatId: string, newChatName: string) => {
        await axios
          .post(BASE_URL(`chat/${chatId}/rename`), { id: chatId, updates: { name: newChatName }, user: get().activeUser })
          .then(() => {
            useBreadCrumbStore.getState().updateBreadcrumb(chatId, newChatName);
            set({ isNameUpdate: true });
            const activeChat = get().activeChat;
            if (activeChat !== undefined && activeChat.treeItemId === chatId) {
              get().setActiveChat({ ...activeChat, name: newChatName });
            }
            get().fetchChatList();
          });
      }),
      /**
       * Updates the sorting type and fetches the updated list of chats and folders.
       *
       * @param {string} newSortType - The new sorting type.
       */
      setSortType: (newSortType: string) => {
        set({ sortType: newSortType });
        get().fetchChatList();
      },
      /**
       * Sets the active folder and fetches the list of chats within the active folder.
       *
       * @param {FolderItem|undefined} newActiveFolder - The new active folder.
       */
      setActiveFolder: (newActiveFolder?: FolderItem) => {
        if (newActiveFolder) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'folder',
            label: newActiveFolder.name,
            href: 'javascript:;',
            onClick: () => {
              get().setActiveChat();
            },
            testId: newActiveFolder.treeItemId
          });
          get().fetchChatList();
        } else {
          const prevFolder = get().activeFolder;
          if (prevFolder) {
            useBreadCrumbStore.getState().removeBreadcrumb(prevFolder.treeItemId);
          }

          get().setIsMatterFilesActive(false);
          get().setIsMattersettingsActive(false);
          get().fetchChatList();
        }
        set({ activeFolder: newActiveFolder });
      },
      /**
       * Sets the active chat.
       *
       * @param {TreeItem|undefined} newActiveChat - The new active chat.
       */
      setActiveChat: (newActiveChat?: TreeItem) => {
        if (newActiveChat) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'message-lines',
            label: newActiveChat.name,
            href: 'javascript:;',
            onClick: () => {},
            testId: newActiveChat.treeItemId
          });
        }
        const prevActiveChat = get().activeChat;
        if (prevActiveChat) {
          useBreadCrumbStore.getState().removeBreadcrumb(prevActiveChat.treeItemId);
        }

        get().setIsMattersettingsActive(false);
        get().setIsMatterFilesActive(false);
        set({ activeChat: newActiveChat });
      },
      /**
       * Sets the user authentication state.
       *
       * @param {UserAuth} userAuth - The user authentication state.
       */
      setUserAuth: (userAuth: UserAuth) => {
        set({ userAuth });
      },
      /**
       * Toggle archived matters view
       */
      toggleArchivedMattersView: () => {
        set({ archivedMattersView: !get().archivedMattersView });

        if (get().archivedMattersView) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'box-archive',
            label: 'Archived Matters',
            href: 'javascript:;',
            onClick: () => {
              get().setActiveFolder(undefined);
            },
            testId: 'archived_matters'
          });
        } else {
          useBreadCrumbStore.getState().removeBreadcrumb('archived_matters');
        }

        get().fetchChatList();
        /** Sets the matter settings active state.
         *
         * @param {boolean} isMatterSettingsActive - The value to be set.
         */
      },
      setIsMattersettingsActive: (isMatterSettingsActive: boolean) => {
        if (isMatterSettingsActive) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'gear',
            label: 'Matter settings',
            href: 'javascript:;',
            onClick: () => {},
            testId: 'matter_settings'
          });
        } else {
          useBreadCrumbStore.getState().removeBreadcrumb('matter_settings');
        }
        set({ isMatterSettingsActive });
      },
      /**
       * Sets the matter settings active state.
       *
       * @param {boolean} isMatterSettingsActive - The value to be set.
       */
      setIsMatterFilesActive: (isMatterFilesActive: boolean) => {
        if (isMatterFilesActive) {
          useBreadCrumbStore.getState().addBreadcrumb({
            icon: 'files',
            label: 'Matter files',
            href: 'javascript:;',
            onClick: () => {},
            testId: 'matter_files'
          });
        } else {
          useBreadCrumbStore.getState().removeBreadcrumb('matter_files');
        }
        set({ isMatterFilesActive });
      },
      refreshActiveFolder: async () => {
        const activeFolder = get().activeFolder;
        if (!activeFolder) return;

        try {
          const res = await axios.get(BASE_URL(Endpoints.LIST), {
            params: {
              sortType: get().sortType,
              parentId: get().activeFolder?.treeItemId,
              onlyArchivedMatters: get().archivedMattersView,
              uid: get().activeUser.id
            }
          });
          
          const chatsAndFolders = res.data;
          if (Array.isArray(chatsAndFolders)) {
            const folderToRefresh = chatsAndFolders.find((item) => item.treeItemId === activeFolder.treeItemId);
            if (folderToRefresh) {
              set({ activeFolder: folderToRefresh });
            }
          } else {
            console.error('API returned unexpected data structure for refreshActiveFolder:', chatsAndFolders);
            useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
          }
        } catch (error) {
          console.error('Error refreshing active folder:', error);
          useMfeContextUtilsStore.getState().triggerAlert(i18nInstance.t('ERROR_MESSAGE'), 'error');
        }
      },
      setIsNameUpdate(isNameUpdateNewValue) {
        set({ isNameUpdate: isNameUpdateNewValue });
      },
      setActiveUser(activeUser: UserProfile) {
        set({ activeUser });
      }
    }),
    { name: 'useChatSidebarStore' }
  )
);
