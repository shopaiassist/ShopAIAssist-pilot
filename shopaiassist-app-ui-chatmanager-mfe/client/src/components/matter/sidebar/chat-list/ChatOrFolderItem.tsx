import { Icon, Text, Tooltip } from '@on/core-components/react';
import { useChatOrFolderItem } from '../../../../hooks/useChatOrFolderItem';
import { SelectableItem, TypewriterText } from '../../../common';
import { TreeItem, FolderItem } from '../../../../@types/sidebar';
import { truncate } from 'lodash';
import { CHAT_OR_FOLDER_NAME_DISPLAY_LENGTH, TYPEWRITER_DELAY } from '../../../../utils/constatnts';

import './ChatOrFolderItem.scss';
import ChatOrFolderMenu from '../../../common/ChatOrFolderMenu';

/**
 * Props for the ChatOrFolderItem component.
 *
 * @interface ChatOrFolderItemProps
 * @property {TreeItem | FolderItem} treeItem - The chat or folder object to display.
 * @property {boolean} [selected] - Indicates whether the item is selected.
 * @property {(item?: TreeItem | FolderItem) => void} onSelect - Callback fired when the item is selected.
 */
interface ChatOrFolderItemProps {
  treeItem: TreeItem | FolderItem;
  selected?: boolean;
  onSelect: (item?: TreeItem | FolderItem) => void;
}

/**
 * Component to display a chat or folder item with context menu options.
 * It uses `SelectableItem` for selectable UI elements, and conditional rendering
 * to switch between chat and folder specific actions. Uses custom hooks for state
 * management and outside click detection.
 *
 * @param {ChatOrFolderItemProps} props - The props for the component.
 * @returns {JSX.Element} The rendered element for either a chat or a folder item.
 */
const ChatOrFolderItem = ({ treeItem, selected, onSelect }: ChatOrFolderItemProps) => {
  const [chatOrFolderItemMenuState, toggleChatOrFolderItemMenu] = useChatOrFolderItem(
    false,
    treeItem.treeItemId,
    treeItem.type
  );

  return (
    <div className="chat-or-folder-item-container">
      <>
        <SelectableItem
          selected={selected}
          onSelect={() => onSelect(treeItem || undefined)}
          dataTestId={`${treeItem?.type}-${treeItem?.treeItemId}-selectable`}
          treeItemId={treeItem.treeItemId}
        >
          <Icon
            appearance={selected ? 'solid' : 'light'}
            size={16}
            iconName={treeItem?.type === 'folder' ? 'folder' : 'message-lines'}
            className="chat-or-folder-icon"
          />
          <Text appearance="body-strong-md" id={`chat-text-${treeItem?.treeItemId}`}>
            <TypewriterText
              text={truncate(treeItem?.name, { length: CHAT_OR_FOLDER_NAME_DISPLAY_LENGTH })}
              delay={TYPEWRITER_DELAY}
            />
          </Text>
          {treeItem && treeItem?.name.length > CHAT_OR_FOLDER_NAME_DISPLAY_LENGTH ? (
            <Tooltip placement="bottom-start" anchor={`chat-text-${treeItem?.treeItemId}`}>
              {treeItem?.name}
            </Tooltip>
          ) : (
            <></>
          )}
        </SelectableItem>
        {treeItem && (
          <ChatOrFolderMenu
            isOpen={chatOrFolderItemMenuState}
            toggleOpen={toggleChatOrFolderItemMenu}
            itemId={treeItem.treeItemId}
            itemType={treeItem.type}
            itemName={treeItem.name}
          />
        )}
      </>
    </div>
  );
};
export default ChatOrFolderItem;
