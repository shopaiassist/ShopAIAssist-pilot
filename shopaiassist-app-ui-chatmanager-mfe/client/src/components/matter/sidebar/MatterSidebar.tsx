import { Text } from '@on/core-components/react';
import MatterSidebarHeader from './header/MatterSidebarHeader';
import ChatList from './chat-list/ChatList';
import { FolderItem, TreeItem } from '../../../@types/sidebar';
import MatterSidebarNavigation from './header/MatterSidebarNavigation';
import useMeasure from 'react-use-measure';
import { animated, useSpring } from '@react-spring/web';
import { useChatSidebarStore } from '../../../store/useChatSidebarStore';
import MatterSidebarFooter from './footer/MatterSidebarFooter';

import './MatterSidebar.scss';
import classNames from 'classnames';
import { ModalType } from '../../../App';
import { useMatters } from '../../../views/Context';

interface MatterSidebarProps {
  isOpen: boolean;
  tree?: (TreeItem | FolderItem)[];
  onNewChat: () => void;
  closeSidebar: () => void;
}

/**
 * A responsive sidebar component for displaying a tree of items such as chats and folders.
 * This sidebar can be toggled to show or hide using an animated transition.
 * The component adapts to changes in the sidebar's width and employs animations for smooth transitions.
 *
 * @param {MatterSidebarProps} props - The properties for the MatterSidebar component.
 * @param {(TreeItem | FolderItem)[]} [props.tree] - Data structure representing the tree items to be displayed in the sidebar.
 * @param {() => void} props.onNewChat - Callback function that triggers when a new chat is initiated.
 * @returns {JSX.Element} The animated sidebar containing a header and a list of chat or folder items.
 *
 * @example
 * <MatterSidebar tree={treeData} onNewChat={handleNewChat} />
 */
const MatterSidebar = ({ isOpen, tree, onNewChat, closeSidebar }: MatterSidebarProps) => {
  const [measureRef, { width }] = useMeasure();
  const springProps = useSpring({
    transform: !isOpen ? `translateX(-${width}px)` : 'translateX(0px)',
    marginRight: !isOpen ? `-${width}px` : '0px'
  });
  const [
    activeFolder,
    setIsMattersettingsActive,
    setIsMatterFilesActive,
    setActiveChat,
    archivedMattersView,
    toggleArchivedMattersView
  ] = useChatSidebarStore((state) => [
    state.activeFolder,
    state.setIsMattersettingsActive,
    state.setIsMatterFilesActive,
    state.setActiveChat,
    state.archivedMattersView,
    state.toggleArchivedMattersView
  ]);
  const { modalType } = useMatters();

  return (
    <animated.div
      style={springProps}
      className={classNames('matter-sidebar-container', {
        'compact-mode': modalType === ModalType.Compact,
        'full-mode': modalType !== ModalType.Compact
      })}
    >
      <div
        ref={measureRef}
        className={classNames('matter-sidebar', {
          'compact-mode': modalType === ModalType.Compact,
          'full-mode': modalType !== ModalType.Compact
        })}
        data-testid="matter-sidebar"
        is-open={isOpen.toString()}
      >
        <MatterSidebarNavigation
          className="matter-sidebar-nav-header"
          isSubsectionView={!!activeFolder || !!archivedMattersView}
          closeSidebar={closeSidebar}
        />
        {activeFolder && (
          <div className="matter-sidebar-folder-name">
            <Text appearance="heading-lg">{activeFolder?.name}</Text>
          </div>
        )}
        {archivedMattersView && (
          <div className="matter-sidebar-folder-name">
            <Text appearance="heading-lg">Archived Matters</Text>
          </div>
        )}
        <MatterSidebarHeader
          className="matter-sidebar-header"
          onNewChat={() => {
            onNewChat();
            if (modalType === ModalType.Compact) {
              closeSidebar();
            }
          }}
          showAddChatButton={!archivedMattersView}
          isFolderView={!!activeFolder}
        />
        <ChatList chatsAndFolders={tree} closeSidebar={closeSidebar} />

        {/* <MatterSidebarFooter
          activeFolder={activeFolder}
          archivedMattersView={archivedMattersView}
          onOpenSettings={() => {
            setIsMattersettingsActive(true), setIsMatterFilesActive(false);
          }}
          onOpenMatterFiles={() => {
            setActiveChat();
            setIsMatterFilesActive(true);
          }}
          toggleArchivedMattersView={toggleArchivedMattersView}
        /> */}
      </div>
    </animated.div>
  );
};

export default MatterSidebar;
