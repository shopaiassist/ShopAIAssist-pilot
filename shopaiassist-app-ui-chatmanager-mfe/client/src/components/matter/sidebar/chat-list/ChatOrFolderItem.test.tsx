import { act, renderWithI18n, snapshot, waitFor } from '../../../../../jest-utils';
import { FolderItem, TreeItem } from '../../../../@types/sidebar';
import { mockTree } from '../../../../utils/test-utils/mockData';
import ChatOrFolderItem from './ChatOrFolderItem';
import { useModalState } from '../../../../store/useModalStore';
import { MatterDialogState, useMatterActionsStore } from '../../../../store/useMatterActionsStore';
import { useChatSidebarStore } from '../../../../store/useChatSidebarStore';

const mockTreeData = mockTree;

const getMockedComponent = (opts: {
  mockOnSelect: (item?: TreeItem | FolderItem) => void;
  mockTreeItem: TreeItem | FolderItem;
}) => {
  return <ChatOrFolderItem treeItem={opts.mockTreeItem} onSelect={opts?.mockOnSelect} />;
};

describe('ChatOrFolderItem Tests', () => {
  const chatSidebarStore = useChatSidebarStore.getState();
  const modalStore = useModalState.getState();
  const matterActionStore = useMatterActionsStore.getState();

  beforeEach(() => {
    useChatSidebarStore.setState(chatSidebarStore, true);
    useModalState.setState(modalStore, true);
    useMatterActionsStore.setState(matterActionStore, true);
  });

  describe('render', () => {
    it('should render', async () => {
      snapshot(getMockedComponent({ mockOnSelect: () => {}, mockTreeItem: mockTreeData[0] }));
    });
  });

  describe('user actions', () => {
    const mockOnSelect = jest.fn();

    const folderActions = async (action: MatterDialogState) => {
      useChatSidebarStore.setState({ chatsAndFolders: [mockTreeData[0]] });
      let buttonDataTestId = '';
      switch (action) {
        react MatterDialogState.ARCHIVE_MATTER:
          buttonDataTestId = 'folder-menu-item-ARCHIVE_FOLDER';
          break;
        react MatterDialogState.DELETE_MATTER:
          buttonDataTestId = 'folder-menu-item-DELETE_MATTER';
          break;
        react MatterDialogState.UPDATE_MATTER:
          buttonDataTestId = 'folder-menu-item-EDIT_FOLDER';
          break;
        default:
          break;
      }
      const component = getMockedComponent({ mockOnSelect, mockTreeItem: mockTreeData[0] });
      const wrapper = renderWithI18n(component);
      const chatOrFolderItemOptionMenuButton = wrapper.getByTestId(
        `${mockTreeData[0].type}-${mockTreeData[0].treeItemId}-option-menu`
      );
      chatOrFolderItemOptionMenuButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      const chatOrFolderItemOpenMenu = wrapper.getByTestId('chat-or-folder-saf-menu');
      expect(chatOrFolderItemOpenMenu).toBeVisible();
      await act(() => waitFor(() => wrapper.rerender(component)));
      const chatDeleteOption = wrapper.getByTestId(buttonDataTestId);
      chatDeleteOption.click();
      expect(useMatterActionsStore.getState().actionItem).toBe(mockTreeData[0]);
      expect(useMatterActionsStore.getState().dialogState).toBe(action);
    };

    it('should call select function when list item is clicked', async () => {
      const component = getMockedComponent({ mockOnSelect, mockTreeItem: mockTreeData[0] });
      const wrapper = renderWithI18n(component);
      const chatOrFolderItem = wrapper.getByTestId(`${mockTreeData[0].type}-${mockTreeData[0].treeItemId}-selectable`);
      chatOrFolderItem.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      expect(mockOnSelect).toHaveBeenCalledWith(mockTreeData[0]);
    });

    it('should open options menu when list item is clicked', async () => {
      const component = getMockedComponent({ mockOnSelect, mockTreeItem: mockTreeData[0] });
      const wrapper = renderWithI18n(component);
      const chatOrFolderItemOptionMenuButton = wrapper.getByTestId(
        `${mockTreeData[0].type}-${mockTreeData[0].treeItemId}-option-menu`
      );
      chatOrFolderItemOptionMenuButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      const chatOrFolderItemOpenMenu = wrapper.getByTestId('chat-or-folder-saf-menu');
      expect(chatOrFolderItemOpenMenu).toBeVisible();
    });

    it('should open options and call to delete chat', async () => {
      useChatSidebarStore.setState({ chatsAndFolders: [mockTreeData[3]] });
      const component = getMockedComponent({ mockOnSelect, mockTreeItem: mockTreeData[3] });
      const wrapper = renderWithI18n(component);
      const chatOrFolderItemOptionMenuButton = wrapper.getByTestId(
        `${mockTreeData[3].type}-${mockTreeData[3].treeItemId}-option-menu`
      );
      chatOrFolderItemOptionMenuButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));
      const chatOrFolderItemOpenMenu = wrapper.getByTestId('chat-or-folder-saf-menu');
      expect(chatOrFolderItemOpenMenu).toBeVisible();
      await act(() => waitFor(() => wrapper.rerender(component)));
      const chatDeleteOption = wrapper.getByTestId('chat-menu-item-DELETE_CHAT');
      chatDeleteOption.click();
      expect(useModalState.getState().actionItem).toBe(mockTreeData[3]);
    });

    it('should open options and call to edit folder', async () => {
      await folderActions(MatterDialogState.UPDATE_MATTER);
    });

    it('should open options and call to archive folder', async () => {
      await folderActions(MatterDialogState.ARCHIVE_MATTER);
    });

    it('should open options and call to delete folder', async () => {
      await folderActions(MatterDialogState.DELETE_MATTER);
    });
  });
});
