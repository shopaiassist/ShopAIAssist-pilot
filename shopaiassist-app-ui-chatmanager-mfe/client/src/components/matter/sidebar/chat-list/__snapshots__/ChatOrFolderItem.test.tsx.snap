// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatOrFolderItem Tests render should render 1`] = `
<div
  class="chat-or-folder-item-container"
>
  <div
    class="selectable-item"
    data-testid="folder-9a4144e0-65ea-4f6c-8cd5-43a0a6d1e634-selectable"
  >
    <div
      appearance="light"
      class="chat-or-folder-icon"
      iconname="folder"
      size="16"
    />
    <span
      appearance="body-strong-md"
      id="chat-text-9a4144e0-65ea-4f6c-8cd5-43a0a6d1e634"
    >
      Test Folder For Created At 2
    </span>
  </div>
  <div
    class="chat-or-folder-menu"
  >
    <div
      class="selectable-item selectable-icon"
      data-testid="folder-9a4144e0-65ea-4f6c-8cd5-43a0a6d1e634-option-menu"
    >
      <div
        iconname="ellipsis-vertical"
        size="24"
      />
    </div>
  </div>
</div>
`;
